samza-core/src/main/java/org/apache/samza/operators/stream/IntermediateStreamInternalImpl.java

```diff
@@ -33,13 +33,13 @@
   private final Function<M, V> msgExtractor;
   private final BiFunction<K, V, M> msgBuilder;
 
-  public IntermediateStreamInternalImpl(StreamGraphImpl graph, StreamSpec streamSpec,
-      Function<M, K> keyExtractor, Function<M, V> msgExtractor, BiFunction<K, V, M> msgBuilder) {
+  public IntermediateStreamInternalImpl(StreamGraphImpl graph, StreamSpec streamSpec, Function<? super M, ? extends K> keyExtractor,
```

Comment:
A little cleaner to do these casts in StreamGraph and leave these types as K, V, M. Both here and InputStreamInternalImpl.